src/serac/physics/utilities/functional/tests/quadrature_data_cuda.cu

```diff
@@ -0,0 +1,396 @@
+
+struct ThreeBytesWrapper {
+  using value_type = ThreeBytes;
+  static_assert(sizeof(value_type) == 3);
+  static constexpr value_type     initial_state = {};
+  __host__ __device__ static void mutate(value_type& v, double other = 0.0)
+  {
+    v.x[0] = static_cast<char>(v.x[0] + 3 + (other * 10));
+    v.x[1] = static_cast<char>(v.x[1] + 2 + (other * 10));
+    v.x[2] = static_cast<char>(v.x[2] + 1 + (other * 10));
+  }
+};
+
+using StateTypes = ::testing::Types<MultiFieldWrapper, IntWrapper, ThreeBytesWrapper>;
+// NOTE: The extra comma is here due a clang issue where the variadic macro param is not provided
+// so instead, we leave it unspecified/empty
+TYPED_TEST_SUITE(QuadratureDataGPUStateManagerTest, StateTypes, );
+
+template <typename wrapper_t>
+struct state_manager_qfunction {
+  template <typename x_t, typename field_t, typename state_t>
+  __host__ __device__ auto operator()(x_t&& /* x */, field_t&& u, state_t&& state)
+  {
+    wrapper_t::mutate(state);
+    return u;
+  }
+};
+
+template <typename wrapper_t>
+struct state_manager_varying_qfunction {
+  template <typename x_t, typename field_t, typename state_t>
+  __host__ __device__ auto operator()(x_t&& x, field_t&& u, state_t&& state)
+  {
+    double norm = 0.0;
+    for (int i = 0; i < x.first_dim; i++) {
+      norm += x[i] * x[i];
+    }
+    wrapper_t::mutate(state, norm);
+    mutated_data[idx++] = state;
+    return u;
+  }
+  DeviceArray<typename wrapper_t::value_type>& mutated_data;
+  int                                          idx = 0;
+};
+
+TYPED_TEST(QuadratureDataGPUStateManagerTest, basic_integrals_state_manager)
+{
+  constexpr int cycle        = 0;
+  const auto    mutated_once = []() {
+    typename TestFixture::value_type result = TestFixture::initial_state;
+    TestFixture::mutate(result);
+    return result;
+  }();
+  const auto mutated_twice = []() {
+    typename TestFixture::value_type result = TestFixture::initial_state;
+    TestFixture::mutate(result);
+    TestFixture::mutate(result);
+    return result;
+  }();
+
+  // First set up the Functional object, run it once to update the state once,
+  // then save it
+  {
+    axom::sidre::DataStore datastore;
+    serac::StateManager::initialize(datastore);
+    // We need to use "this->" explicitly because we are in a derived class template
+    serac::StateManager::setMesh(std::move(this->default_mesh));
+    // Can't use auto& here because we're in a template context
+    serac::QuadratureData<typename TestFixture::value_type>& qdata =
+        serac::StateManager::newQuadratureData<typename TestFixture::value_type>("test_data", this->p);
+    qdata = TestFixture::initial_state;
+
+    this->residual->AddDomainIntegral(Dimension<TestFixture::dim>{}, state_manager_qfunction<TestFixture>{},
+                                      *this->mesh, qdata);
+    // If we run through it one time...
+    mfem::Vector U(this->festate->space().TrueVSize());
+    (*this->residual)(U);
```

Comment:
I think you can avoid prefixing members by `this->` if you put an appropriate `using` statement in the class, e.g.

```cpp
 template <typename T>
 class QuadratureDataGPUStateManagerTest : public QuadratureDataGPUTest {
 public:

   using QuadratureDataGPUTest::residual; /// <------

   using value_type                              = typename T::value_type;
   static constexpr value_type     initial_state = T::initial_state;
   __host__ __device__ static void mutate(value_type& v, double other = 0.0) { T::mutate(v, other); }
 };
```

but I'm not sure if this is much better, it just moves some of the weirdness to the class definition, rather than where it is used.